engine/src/main/java/org/hibernate/validator/internal/cfg/context/PropertyConstraintMappingContextImpl.java

```diff
@@ -17,32 +17,50 @@
 import org.hibernate.validator.internal.metadata.core.ConstraintHelper;
 import org.hibernate.validator.internal.metadata.descriptor.ConstraintDescriptorImpl.ConstraintType;
 import org.hibernate.validator.internal.metadata.location.ConstraintLocation;
-import org.hibernate.validator.internal.metadata.raw.ConfigurationSource;
 import org.hibernate.validator.internal.metadata.raw.ConstrainedElement;
-import org.hibernate.validator.internal.metadata.raw.ConstrainedExecutable;
-import org.hibernate.validator.internal.metadata.raw.ConstrainedField;
-import org.hibernate.validator.internal.properties.Callable;
 import org.hibernate.validator.internal.properties.Property;
-import org.hibernate.validator.internal.properties.javabean.JavaBeanField;
 import org.hibernate.validator.internal.util.TypeResolutionHelper;
+import org.hibernate.validator.internal.util.logging.Log;
+import org.hibernate.validator.internal.util.logging.LoggerFactory;
 
 /**
  * Constraint mapping creational context which allows to configure the constraints for one bean property.
  *
  * @author Hardy Ferentschik
  * @author Gunnar Morling
  * @author Kevin Pollet &lt;kevin.pollet@serli.com&gt; (C) 2011 SERLI
+ * @author Marko Bekhta
  */
-final class PropertyConstraintMappingContextImpl
+abstract class PropertyConstraintMappingContextImpl
 		extends CascadableConstraintMappingContextImplBase<PropertyConstraintMappingContext>
 		implements PropertyConstraintMappingContext {
 
+	private static final Log LOG = LoggerFactory.make( MethodHandles.lookup() );
+
 	private final TypeConstraintMappingContextImpl<?> typeContext;
 
 	// either Field or Method
 	private final Property property;
 	private final ConstraintLocation location;
 
+	static PropertyConstraintMappingContextImpl context(ElementType elementType, TypeConstraintMappingContextImpl<?> typeContext, Property property) {
+		if ( elementType == ElementType.FIELD ) {
+			return new FieldPropertyConstraintMappingContextImpl(
+					typeContext,
+					property
+			);
+		}
+		else if ( elementType == ElementType.METHOD ) {
+			return new GetterPropertyConstraintMappingContextImpl(
+					typeContext,
+					property
+			);
+		}
+		else {
+			throw LOG.getUnexpectedElementType( elementType, ElementType.FIELD, ElementType.METHOD );
+		}
+	}
```

Comment:
I wonder if we should deprecate `property(String property, ElementType type)` and introduce `field()` and `getter()`. That would be more in line with what we do in the XML.

Then I would instantiate directly `FieldPropertyConstraintMappingContextImpl` and `GetterPropertyConstraintMappingContextImpl` in those methods and disambiguate the types and throw the exception in the deprecated `property()`.

That if we take the road of having a specific DSL for JSON. My opinion is that it will be better. WDYT?

Reply:
Applied the change in the last commit. As for the JSON and all that "free-form validation" I'd say that as a lot of stuff there would be different and probably tagged as `@Incubating` we'd better separate these things. We already know that cascading would be different and for properties we need to pass a property type.